Source/Dafny/Compiler.cs

```diff
@@ -697,13 +698,29 @@ public enum ResolvedUnaryOp { BoolNot, BitwiseNot, Cardinality }
             }
             var classIsExtern = false;
             if (include) {
-              classIsExtern = !DafnyOptions.O.DisallowExterns && Attributes.Contains(cl.Attributes, "extern");
+              classIsExtern = !DafnyOptions.O.DisallowExterns && Attributes.Contains(cl.Attributes, "extern") || cl.IsDefaultClass && Attributes.Contains(cl.Module.Attributes, "extern");
               if (classIsExtern && cl.Members.TrueForAll(member => member.IsGhost || Attributes.Contains(member.Attributes, "extern"))) {
                 include = false;
               }
             }
+            // If the class is declared {:extern} but no members are, then it's
+            // not *really* an extern class, it's just a Dafny class with a
+            // specified externally-visible name.  Thus we will call CreateClass
+            // with isExtern set to false in this case.  This is particularly
+            // important for Java, where the user must provide an implementation
+            // for every extern class, and if there are no extern methods to
+            // implement, it would do nothing but extend the generated base
+            // class.  This case is less niche than it sounds, since an extern
+            // declaration on a module is useful for calling Dafny code from
+            // extern code even if the module has no extern methods, and we
+            // treat the default class of an extern module as an extern class.
```

Comment:
This doesn't quite sit right with me, mostly because the precise abstract meaning of `{:extern}` hasn't been documented AFAICT, looking through the release notes or reference manual. I would really prefer it if we could have separate language features for "externally implemented" vs. "can be referenced from external code". I worry we'll hit contradictions in some cases trying to make one attribute cover both (see #461 and #463).

I believe the most accurate interpretation of `{:extern}` right now is "externally implemented", so using it just on a class to make the second use case possible should be considered, at best, relying on undefined behaviour. You shouldn't be able to assume that a method without `{:extern}` has an un-mangled name, or even exists (if the compiler starts to do more optimizations). 

Would you consider introducing a new attribute for the second case? Unfortunately my personal preference would probably be using `{:native}` for "externally implemented" and `{:extern}` for "externally reference-able".

Reply:
That makes sense to me, though I don't think introducing the attribute is in scope for the PR.  I'm inclined to take out the comment (and the special case it refers to) and file an issue about adding a new attribute.

Reply:
Sure, that seems reasonable, if you're able to work around the lack of that special case without too much pain. The consequence is just creating a pointless concrete Java subclass, correct?

Reply:
Yup.  Really not too bad, as workarounds go.  I'll take out the special case.